### ISSUE TITLE
Integrate Static Code Analysis

### ISSUE THREAD

**Issue by skunert** [stratosphere/stratosphere#536]
In case no annotations where made, a static code analysis should be performed to detect the write, read and copysets.

**Comment by skunert** [comment]
There is a big problem with the current solution. I integrated the static code analysis into the FunctionAnnotation class.

I am having problems detecting the right copysets when a record is taken from input and handed to the collector directly. In this case, any fields that are not touched should appear in the copyset. Therefore, I need to know how many fields a record has to add them to the copyset. I think it is not possible to detect the number of fields one record has. Without this information, the analyzer can make no assumption about any operator that hands an input record to the collector.

**Comment by StephanEwen** [comment]
Can you have a look at the ImplicitlyForwardingSemanticProperties (or so) ? It is a subclass used by the FunctionAnnotations for exactly that case (ConstantExcept cases)

**Comment by rmetzger** [comment]
This was the original issue for the task: https://github.com/stratosphere/stratosphere/issues/378

**Comment by skunert** [comment]
In the current state the analyzer is working quite well, only very few tests fail. The test  eu.stratosphere.test.localDistributed.PackagedProgramEndToEndTest raises a exception because soot is unable to find the class that should be analyzed. The exception is raised in the FunctionAnnotation class. I am confused why the class can not be found since I think the class is loaded before I hand it to the analyzer.

**Comment by skunert** [comment]
I was able to fix this problem by getting the class loader from the udf and extracting the classpath. This classpath is then appended to the soot classpath.

**Comment by skunert** [comment]
It turned out that this solution works fine for testing with eclipse. With maven-verify there are still classpath issues.

**Comment by StephanEwen** [comment]
Relative paths are different in Eclipse and maven. Did you try to convert
url (to uri) to absolute file path? That would most likely work.

**Comment by skunert** [comment]
Yeah I did that, but it does not work. At the moment, i am building a new classpath from the system classloader, the classloader of the udf, the path that is returned from udf.getProtectionDomain().getCodeSource().getLocation().getPath() and the previous soot path. I have no additional ideas on how to get the classes that are needed at the moment.

**Comment by StephanEwen** [comment]
Have you tried to augment the classpath that works in Eclipse with the
System.getProperty() for the current working directory?

**Comment by skunert** [comment]
Yeah, the current working directory is in the path. I print the classpath and in the maven tests it is entirely different than the classpath that is printed from eclipse. In Eclipse, a completely different classpath is build from the same commands. I think the problem is related to the maven surefire testing plugin. The classpath always contains the standard libraries and /stratosphere/stratosphere-tests/target/surefire/surefirebooter8229394731766549075.jar. But other entries are missing that are available in eclipse.
